src/compiler/crystal/semantic/type_guess_visitor.cr

```diff
@@ -800,6 +803,20 @@ module Crystal
       if arg = args_hash[node.name]?
         # If the argument has a restriction, guess the type from it
         if restriction = arg.restriction
+          # It is for something like `def foo(*@foo : *T)`.
+          if @splat == arg
```

Comment:
I'd do `@splat.same?(arg)`

Reply:
Thank you. I think so too. I'll fix it later.